### PR TITLE
[ONNX] Im2col export

### PR DESCRIPTION
**Pull request by neginraoof**
Added im2col to opset 11.
This symbolic is used to export torch.nn.Unfold

### CODE REVIEW

**Comment by neginraoof** [comment]
cc @BowenBao @lara-hdr
Please review

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 1]
```diff
@@ -2135,6 +2134,15 @@ def test_replication_pad(self):
         x = torch.randn(2, 2, 4, 4)
         self.run_test(model, x)
 
+    @skipIfUnsupportedMinOpsetVersion(11)
+    def test_im2col(self):
+        class Unfold(torch.nn.Module):
+            def forward(self, input):
+                return torch.nn.functional.unfold(input, kernel_size=(10, 15), dilation=(2, 3), padding=5, stride=3)
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py]: Consider adding more test cases to cover edge cases.

File: torch/onnx/symbolic_opset11.py [hunk 2]
```diff
@@ -465,3 +465,78 @@ def __lshift_(g, self, other):
...
+
+    blocks_row_indices = _get_im2col_indices_along_dim(g, input_h, kernel_h, dilation_h, padding_h, stride_h)
+    blocks_col_indices = _get_im2col_indices_along_dim(g, input_w, kernel_w, dilation_w, padding_w, stride_w)
+
+    output_shape = _get_im2col_output_shape(g, input, kernel_h, kernel_w)
+    padded_input = _get_im2col_padded_input(g, input, padding_h, padding_w)
+
+    output = g.op("Gather", padded_input, blocks_row_indices, axis_i=2)
+    output = g.op("Gather", output, blocks_col_indices, axis_i=4)
+    output = g.op("Transpose", output, perm_i=[0, 1, 2, 4, 3, 5])
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset11.py]: Consider adding more comments explaining what is happening here ... Like what the shape of output is after each Gather operation, and what each axis represents.

File: torch/onnx/symbolic_opset11.py [hunk 3]
```diff
@@ -465,3 +465,78 @@ def __lshift_(g, self, other):
...
+
+    # Apply dilation on kernel and find its indices along dim d
+    kernel_grid = g.op("Range",
+                       g.op("Constant", value_t=torch.tensor(0, dtype=torch.int64)),
+                       g.op("Constant", value_t=torch.tensor(kernel_size_d * dilation_d)),
+                       g.op("Constant", value_t=torch.tensor(dilation_d)))
+
+    # Broadcast and add kernel staring positions (indices) with
+    # kernel_grid along dim d, to get block indices along dim d
+    blocks_d_indices = g.op('Reshape', blocks_d_indices, g.op('Constant', value_t=torch.tensor([1, -1])))
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset11.py]: nit: could use unsqueeze as well.

File: torch/onnx/symbolic_opset11.py [hunk 4]
```diff
@@ -465,3 +465,78 @@ def __lshift_(g, self, other):
...
+
+    blocks_d = g.op("Add", input_d, g.op("Constant", value_t=torch.tensor(padding_d * 2)))
+    blocks_d = g.op("Sub", blocks_d, g.op("Constant", value_t=torch.tensor(dilation_d * (kernel_size_d - 1))))
+
+    # Stride kernel over input and find starting indices along dim d
+    blocks_d_indices = g.op("Range", g.op("Constant", value_t=torch.tensor(0)),
+                            blocks_d, g.op("Constant", value_t=torch.tensor(stride_d)))
+
+    # Apply dilation on kernel and find its indices along dim d
+    kernel_grid = g.op("Range",
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset11.py]: this thing looks completely constant, can be constructed directly without using Range.

**Review by BowenBao — COMMENTED** [review]
Looks good, the code is very clean. Just some minor comments.

File: torch/onnx/symbolic_opset11.py [hunk 5]
```diff
@@ -514,6 +515,105 @@ def __lshift_(g, self, other):
...
+
+    blocks_d = g.op("Add", input_d, g.op("Constant", value_t=torch.tensor(padding_d * 2)))
+    blocks_d = g.op("Sub", blocks_d, g.op("Constant", value_t=torch.tensor(dilation_d * (kernel_size_d - 1))))
+
+    # Stride kernel over input and find starting indices along dim d
+    blocks_d_indices = g.op("Range", g.op("Constant", value_t=torch.tensor(0)),
+                            blocks_d, g.op("Constant", value_t=torch.tensor(stride_d)))
+
+    # Apply dilation on kernel and find its indices along dim d
+    # kernel_grid = g.op("Range",
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset11.py]: nit: remove commented code

**Review by BowenBao — APPROVED** [review]
LGTM

**Comment by neginraoof** [comment]
@houseroad PR is approved and ready. Could you please take a look?

**Review by houseroad — APPROVED** [review]
LGTM
